sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -278,21 +280,127 @@ public void setResolveFieldsOperation(Operation operation) {
+     * @throws ConversionException If castsIsRequired is true and the type doesn't match and upcasting is not possible.
+     */
+    public static Symbol castIfNeeded(Symbol symbolToCast, DataType targetType, boolean castIsRequired) {
         DataType sourceType = symbolToCast.valueType();
-        if (sourceType.equals(targetType)) {
-            return symbolToCast;
+        if (!symbolToCast.valueType().equals(targetType)) {
+            boolean containsField = SymbolVisitors.any(symbol -> symbol instanceof Field, symbolToCast);
+            if ((symbolToCast.valueType().id() == UndefinedType.ID || !containsField) &&
+                    (sourceType.isConvertableTo(targetType) || canBeCasted(sourceType, targetType, symbolToCast))) {
+                return cast(symbolToCast, targetType, false);
+            } else if (castIsRequired) {
+                throw new ConversionException(symbolToCast, targetType);
+            }
+        }
+        return symbolToCast;
+    }
+
+    /**
+     * Check if a Symbol can be casted without losing entropy.
+     * @param sourceType The symbol type to cast.
+     * @param targetType The target type to cast to.
+     * @param symbolToCast The symbol to cast. If provided enables to check for Literals
+     *                     which may be downcasted.
+     * @return True if the cast can be performed, false otherwise.
+     */
+    private static boolean canBeCasted(DataType sourceType, DataType targetType, @Nullable Symbol symbolToCast) {
+        if (targetType.id() == UndefinedType.ID) {
+            return false;
         }
-        if (sourceType.isConvertableTo(targetType)) {
-            // cast further below doesn't always fail because it might be lazy as it wraps functions/references inside a cast function.
-            // -> Need to check isConvertableTo to fail eagerly if the cast won't work.
-            return cast(symbolToCast, targetType, false);
+        if (DataTypes.isCollectionType(sourceType) && DataTypes.isCollectionType(targetType)) {
+            DataType<?> sourceInnerType = ((CollectionType) sourceType).innerType();
+            DataType<?> targetInnerType = ((CollectionType) targetType).innerType();
+            return canBeCasted(sourceInnerType, targetInnerType, symbolToCast);
         }
-        throw new ConversionException(symbolToCast, targetType);
+        return checkForSpecialTypeHandling(symbolToCast, targetType) ||
+                                    targetType.precedes(sourceType) ||
+                                    sourceType.id() == UndefinedType.ID;
     }
 
+    /**
+     * Indicates whether it is safe to cast to a type which would requires upcasting.
+     * @param sourceSymbol The Symbol to cast to the target type.
+     * @param targetType The target type to cast to.
+     * @return True if it is safe to cast the symbol to the target type.
+     */
+    private static boolean checkForSpecialTypeHandling(Symbol sourceSymbol, DataType targetType) {
```

Comment:
Maybe add nullable annotation for sourceSymbol

Reply:
Not sure if that makes sense here. The null check is just to avoid NPE, we don't want to encourage passing null here.

Reply:
`canBeCasted` has a `Nullable symbolToCast` - which is used in the call to `checkForSpecialTypeHandling` - that's why I thought it would make sense.